x-pack/filebeat/input/gcs/input.go

```diff
@@ -0,0 +1,161 @@
+	inputName string = "gcs"
+)
+
+func Plugin(log *logp.Logger, store cursor.StateStore) v2.Plugin {
+	return v2.Plugin{
+		Name:       inputName,
+		Stability:  feature.Experimental,
+		Deprecated: false,
+		Info:       "Google Cloud Storage",
+		Doc:        "Collect logs from Google Cloud Storage Service",
+		Manager: &cursor.InputManager{
+			Logger:     log,
+			StateStore: store,
+			Type:       inputName,
+			Configure:  configure,
+		},
+	}
+}
+
+func configure(cfg *conf.C) ([]cursor.Source, cursor.Input, error) {
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, nil, err
+	}
+
+	var sources []cursor.Source
+	for _, b := range config.Buckets {
+		bucket := tryOverrideOrDefault(config, b)
+		sources = append(sources, &types.Source{
+			ProjectId:     config.ProjectId,
+			BucketName:    bucket.Name,
+			BucketTimeOut: *bucket.BucketTimeOut,
+			MaxWorkers:    *bucket.MaxWorkers,
+			Poll:          *bucket.Poll,
+			PollInterval:  *bucket.PollInterval,
+		})
+	}
+
+	return sources, &gcsInput{config: config}, nil
+}
+
+// tryOverrideOrDefault, overrides global values with local
+// bucket level values if present. If both global & local values
+// are absent, assigns default values
+func tryOverrideOrDefault(cfg config, b bucket) bucket {
+	if b.MaxWorkers == nil && cfg.MaxWorkers != nil {
+		b.MaxWorkers = cfg.MaxWorkers
+	} else if b.MaxWorkers == nil && cfg.MaxWorkers == nil {
+		workers := 1
+		b.MaxWorkers = &workers
+	}
+
+	if b.Poll == nil && cfg.Poll != nil {
+		b.Poll = cfg.Poll
+	} else if b.Poll == nil && cfg.Poll == nil {
+		poll := false
+		b.Poll = &poll
+	}
+
+	if b.PollInterval == nil && cfg.PollInterval != nil {
+		b.PollInterval = cfg.PollInterval
+	} else if b.PollInterval == nil && cfg.PollInterval == nil {
+		interval := time.Second * 300
+		b.PollInterval = &interval
+	}
+
+	if b.BucketTimeOut == nil && cfg.BucketTimeOut != nil {
+		b.BucketTimeOut = cfg.BucketTimeOut
+	} else if b.BucketTimeOut == nil && cfg.BucketTimeOut == nil {
+		timeout := time.Second * 50
+		b.BucketTimeOut = &timeout
+	}
+
+	return b
+}
+
+func (input *gcsInput) Name() string {
+	return inputName
+}
+
+func (input *gcsInput) Test(src cursor.Source, ctx v2.TestContext) error {
+	return nil
+}
+
+func (input *gcsInput) Run(inputCtx v2.Context, src cursor.Source, cursor cursor.Cursor, publisher cursor.Publisher) error {
+	var cp *state.Checkpoint
+	st := state.NewState()
+	currentSource := src.(*types.Source)
+
+	log := inputCtx.Logger.With("project_id", currentSource.ProjectId).With("bucket", currentSource.BucketName)
+	log.Infof("Running google cloud storage for project: %s", input.config.ProjectId)
+
+	if !cursor.IsNew() {
+		if err := cursor.Unpack(&cp); err != nil {
+			return err
+		}
+
+		st.SetCheckpoint(cp)
+	}
+
+	ctx, cancelInputCtx := context.WithCancel(context.Background())
+	go func() {
+		defer cancelInputCtx()
+		select {
+		case <-inputCtx.Cancelation.Done():
+		case <-ctx.Done():
+		}
+	}()
```

Comment:
ISTM that `<-ctx.Done()` will never fire. Am I missing something? Could this select not be rewritten as
```suggestion
	go func() {
		<-inputCtx.Cancelation.Done()
		cancel()
	}()
```
Note also there is no need for a defer.

Reply:
ya , understood .. thanks for the clarification  ... this is the code used in most other inputs .. and since I copied the skeleton of those inputs initially, these have remained. Will update them.